docs/AQL/grammar/AqlParser.g4

```diff
@@ -0,0 +1,234 @@
+fromClause
+    : FROM fromExpr
+	;
+
+whereClause
+    : WHERE whereExpr
+    ;
+
+orderByClause
+    : ORDER BY orderByExpr (COMMA orderByExpr)*
+    ;
+
+limitClause
+    : LIMIT limit=NN_INTEGER (OFFSET offset=NN_INTEGER)?
+    ;
+
+
+selectExpr
+	: columnVar (AS IDENTIFIER)?
+	;
+
+fromExpr
+    : containsExpr
+    ;
+
+whereExpr
+    : NOT? identifiedExpr
+    | whereExpr AND whereExpr
+    | whereExpr OR whereExpr
+    | OPEN_PAR whereExpr CLOSE_PAR
+    ;
+
+orderByExpr
+	: identifiedPath order=(DESCENDING|DESC|ASCENDING|ASC)?
+	;
+
+
+columnVar
+	: identifiedPath
+	| aggregateFunctionCall
+	| functionCall
+	;
+
+containsExpr
+    : classExprOperand (CONTAINS containsExpr)?
+    | containsExpr AND containsExpr
+    | containsExpr OR containsExpr
+    | OPEN_PAR containsExpr CLOSE_PAR
+    ;
+
+identifiedExpr
+    : EXISTS identifiedPath
+    | identifiedPath COMPARISON_OPERATOR identifiedOperand
+    | functionCall COMPARISON_OPERATOR identifiedOperand
+    | identifiedPath LIKE likeOperand
+    | identifiedPath MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY
+    ;
+
+identifiedOperand
+ 	: primitive
+ 	| PARAMETER
+ 	| identifiedPath
+ 	| functionCall
+ 	;
+
+identifiedPath
+    : IDENTIFIER pathPredicate? (SLASH objectPath)?
+    ;
+
+pathPredicate
+ 	: OPEN_BRACKET (standardPredicate | archetypePredicate | nodePredicate) CLOSE_BRACKET
+ 	;
+
+standardPredicate
+    : objectPath COMPARISON_OPERATOR pathPredicateOperand
+    ;
+
+archetypePredicate
+    : ARCHETYPEID
+    | PARAMETER
+    ;
+
+nodePredicate
+    : NODEID (COMMA (STRING|PARAMETER))?
+    | ARCHETYPEID (COMMA (STRING|PARAMETER))?
+    | PARAMETER
+    | objectPath COMPARISON_OPERATOR pathPredicateOperand
+    | objectPath MATCHES REGEXPATTERN
+    | nodePredicate AND nodePredicate
+    | nodePredicate OR nodePredicate
+    ;
+
+pathPredicateOperand
+    : primitive
+ 	| objectPath
+ 	| PARAMETER
+ 	;
+
+
+objectPath
+ 	: pathPart (SLASH pathPart)*
+ 	;
+pathPart
+ 	: IDENTIFIER pathPredicate?
+ 	;
+
+likeOperand
+    : STRING
+    | PARAMETER
+    ;
+matchesOperand
+ 	: valueListItem (COMMA valueListItem)*
+ 	| terminologyFunction
+ 	| URIVALUE
+ 	;
+valueListItem
+ 	: primitive
+ 	| PARAMETER
+ 	| terminologyFunction
+ 	;
+
+primitive
+    : STRING
+    | INTEGER
+    | FLOAT
+    | DOUBLE
+    | DATE
+    | BOOLEAN
+    ;
+
+
+
+classExprOperand
+	: IDENTIFIER IDENTIFIER? // RM_TYPE_NAME variable
+    | archetypedClassExpr
+    | versionedClassExpr
+	| versionClassExpr
+	;
+
+// RM_TYPE_NAME [archetype_id]
+// RM_TYPE_NAME variable [archetype_id]
+archetypedClassExpr
+ 	: IDENTIFIER IDENTIFIER? archetypePredicate
+ 	;
+versionedClassExpr
+ 	: VERSIONED_OBJECT IDENTIFIER? (OPEN_BRACKET standardPredicate OPEN_BRACKET)?
+ 	;
+versionClassExpr
+ 	: VERSION IDENTIFIER? (OPEN_BRACKET (standardPredicate|versionPredicate) OPEN_BRACKET)?
+ 	;
+versionPredicate
+ 	: LATEST_VERSION
+ 	| ALL_VERSIONS
+ 	;
+
+functionCall
+    : terminologyFunction
+    | function
+    ;
+
+function
```

Comment:
I assume this is intended as an 'extension point' for vendor-specific functions, since there aren't any functions besides `TERMINOLOGY` and the aggregate functions defined in the specification? If so, I would leave it out for now. I have looked into this, and there are ways to override parts of a grammar in ANTLR anyway, if implementers want to add or modify rules.

For example, it is possible to import another parser grammar into this one using the `import` keyword, and override the `functionCall` rule to add more options, or add things in other places. This is less constraining, and maybe less confusing, as functions are defined here now, but not in the non-grammar specification.

The pattern here also doesn't support NOW, CURRENT-DATE or CURRENT-DATE-TIME (or other zero-argument functions), as at least one argument is mandatory.

Reply:
There will be more functions, just that I did not worked on their specification yet - see https://openehr.atlassian.net/browse/SPECQUERY-28

This place on parser is indeed a placeholder for these functions as well as a away of extending them on implementation side. Your proposal with `import` sounds even better, but I would like a way of extending this `functionCall` list instead of overwriting, perhaps by overwriting `function` (which is a generic function) or by adding a new dedicated parser rule for this purpose?

Reply:
is this `import` only for parser grammar (or also lexer) ?

Reply:
Using this as a placeholder for changes to the specification that will be included in this release is fine. I just doubt that providing an extension point for vendor-specific extensions is much use. Also keep in mind that implementers might need to modify the grammar a bit for their implementations anyway (but ofcourse stay functionally equivalent). 

As far as importing goes, there are basically 3 types of grammars: combined grammars, recognizable by `grammar <GrammarName>` as their first line, lexer grammars, recognizable by `lexer grammar <LexerGrammarName>` as their first line, and parser grammars, recognizable by `parser grammar <ParserGrammarName>` as their first line. Now, lexer grammars can import other lexer grammars, parser grammars can import other parser grammars, and combined grammars can import both. 

There is a way to work around the limitation on importing lexer grammars in parser grammars using the `tokenVocab` option you also used in AQLParser.g.